### ISSUE TITLE
Set `syntax on`

### ISSUE THREAD

**Issue by trusktr** [vscode-neovim/vscode-neovim#519]
Comment formatting in Neovim is really good. How can I let Neovim handle that, instead of VS Code? When I press `gq` after selecting a `/** */` comment, it formats awkwardly while in VS Code.

**Comment by theol0403** [comment]
Are you sure you want the `=` operator to be delegated, not `gq`, which I don't think this plugin touches?

Here is where the plugin overrides `=`.

https://github.com/asvetliakov/vscode-neovim/blob/532d4b8809f68fd90b9f4c18484a66ec158441d3/vim/vscode-code-actions.vim#L61

To unmap that, the syntax will be something along the lines of `xnoremap <expr> = =` or unmap, I'm not super familiar with the binding syntax.

Also, maybe you can just get the vscode formatting working properly? It's quite good, maybe you just don't have a good formatter selected for the file type.

**Comment by David-Else** [comment]
This might improve comments for you using visual mode:
```
if exists('g:vscode')

  " call VSCodeCommentary on the visual selection, and then re-select that visual selection with gv
  xmap <C-/> <Plug>VSCodeCommentarygv
  nmap <C-/> <Plug>VSCodeCommentaryLine

else
```

**Comment by trusktr** [comment]
More info:

I'm not sure what the difference is yet, but when I am in plain Neovim and I select a comment and hit `gq`, the expected happens based on my `formatoptions`:

https://user-images.githubusercontent.com/297678/104112797-f2b69480-52a7-11eb-9861-b805bab6961c.mp4

This is what happens in VS Code:

https://user-images.githubusercontent.com/297678/104112813-1679da80-52a8-11eb-9874-0b79ef575695.mp4

If I can just have plain native Neovim handling then I'd be good. I'd rather not try to make new mappings to new things like `VSCodeCommentarygv` if I don't have to, so that I can have a more consistent experience inside or outside VS Code.

**Comment by asvetliakov** [comment]
Curious, we don't touch `gq`, but just confirmed - this happens for me too. I think it's some auto-indenting vscode thing trying to indent when vscode receives new lines from neovim

**Comment by asvetliakov** [comment]
@trusktr What is your filetype value in neovim and vscode?

because: `set filetype=javascript` works normally while `set filetype=typescript` doesn't

**Comment by asvetliakov** [comment]
looks like it's some setting on neovim side. `set indentexpr=""` works for `filetype=typescript` too. Curious it works in terminal too

**Comment by asvetliakov** [comment]
Reproduced in raw neovim:
```
set filetype=typescript
syntax off
```

Syntax is explicitly disabled here:
https://github.com/asvetliakov/vscode-neovim/blob/532d4b8809f68fd90b9f4c18484a66ec158441d3/vim/vscode-options.vim#L76

The reason here since we don't need syntax HL (done by vscode) so we can keep it disabled for perf reason. Such shame it breaks indentation at the same time

**Comment by asvetliakov** [comment]
if anyone knows a way to keep `syntax on` and disable all syntax HL at the same time, sharing it would be very appreciated.

**Comment by theol0403** [comment]
@justinmk is there a way to keep syntax on but disable syntax HL at the same time? Another advantage of this is that we could allow all syntax highlighting groups to be shown in vscode by default (instead of blocking most groups), while not actually having neovim syntax highlighting shown in vscode (which is also bad for performance).

The advantage of this would be that plugins that use existing highlight groups (like vim-sandwich or quick-scope) would work out of the box with no workarounds. For example:

```
  highlight default link OperatorSandwichBuns   IncSearch
  highlight default link OperatorSandwichAdd    DiffAdd
```

**Comment by justinmk** [comment]
> Syntax is explicitly disabled here: https://github.com/asvetliakov/vscode-neovim/blob/532d4b8809f68fd90b9f4c18484a66ec158441d3/vim/vscode-options.vim#L76
> 
> The reason here we don't need syntax HL (done by vscode) so we keep it disabled for perf reasons (and to make processing vim HL groups much easier). Such shame it breaks indentation at the same time

Performance on which side? On the Nvim side, "keep syntax on but disable HL" doesn't help because syntax is the expensive part, not HL.

What is the problem with "processing vim HL groups"?

**Comment by theol0403** [comment]
@justinmk performance on vscode side. Since vscode already does syntax highlighting, we don't want to use nvim's highlighting. If we enable syntax, vscode would be spammed by highlight groups that require expensive vscode decorations to render. The current solution is both to block vim syntax hl groups and disable syntax. 

I don't think the performance hit on nvim is relevant here - nvim already runs magnitudes faster than vscode. However, vscode being spammed by highlight groups from syntax hl would slow down the extension and communication. If vscode doesn't ignore these groups, performance will be really bad because of decorations, but I think even ignoring them has a penalty.

TLDR; we don't use vim syntax groups, and nvim constantly sending them from `syntax on` is expensive, probably even if ignored. However, the goal is that we don't need to ignore them, so that we can support plugins using hl links, and also have other advantages of `syntax on` (without hl).

**Comment by justinmk** [comment]
So don't enable the `ext_hlstate` UI option? Is vscode-neovim using `ext_linegrid`? From `:help ui-option`:

```
By default, Nvim sends |ui-global| and |ui-grid-old| events (for backwards
compatibility); these suffice to implement a terminal-like interface. However
the new |ui-linegrid| represents text more efficiently (especially highlighted
text), and allows UI capabilities requiring multiple grids. New UIs should
implement |ui-linegrid| instead of |ui-grid-old|.
```

**Comment by theol0403** [comment]
@justinmk yes, vscode-neovim is using `ext_linegrid`.

I think you misunderstand what I'm trying to say.

Reasons we want highlights (`ext_hlstate`):
- plugin highlights
	- highlightyank
	- sneak/lightspeed/leap/clever-f
	- surround/sandwich/quick-scope
	- any other plugin that uses highlights to decorate text
- visual selection highlight (currently) 
- vim help syntax hl 
- search highlight (incsearch and n/N)

Reasons we want `syntax on`:
- fix syntax-aware comment formatting (this issue)

Problems with having `syntax on`:
- spamming of highlight groups to vscode-neovim (or so I'd assume)
- required to ignore all syntax highlight groups in vscode-neovim

Reasons we don't want to block any highlight groups:
- so we don't need to patch the highlight groups of plugins that alias to syntax groups

The solution to all this would be if we can have `syntax on` (to solve this issue), while having syntax hl off (to avoid the need to block groups and improve vscode performance).

**Comment by theol0403** [comment]
I've stolen this issue. This is now about how we can set `syntax on` without spamming highlight groups.
